config/config.go

```diff
@@ -39,3 +41,92 @@ func DetermineUseLiveData(logger golog.Logger, liveData *bool, sensors []string)
 	}
 	return useLiveData, nil
 }
+
+// AttrConfig describes how to configure the SLAM service.
+type AttrConfig struct {
+	Sensors             []string          `json:"sensors"`
+	ConfigParams        map[string]string `json:"config_params"`
+	DataDirectory       string            `json:"data_dir"`
+	UseLiveData         *bool             `json:"use_live_data"`
+	DataRateMsec        int               `json:"data_rate_msec"`
+	MapRateSec          *int              `json:"map_rate_sec"`
+	Port                string            `json:"port"`
+	DeleteProcessedData *bool             `json:"delete_processed_data"`
+	Dev                 bool              `json:"dev"`
+}
+
+// NewAttrConfig creates a SLAM config from a service config.
+func NewAttrConfig(cfg config.Service) (*AttrConfig, error) {
+	attrCfg := &AttrConfig{}
+
+	_, err := config.TransformAttributeMapToStruct(attrCfg, cfg.Attributes)
+	if err != nil {
+		return &AttrConfig{}, newError(err.Error())
+	}
+
+	// This temporary value will be replaced once we are using rdk's validation
+	_, err = attrCfg.Validate("services.slam.attributes.fake")
+	if err != nil {
+		return &AttrConfig{}, newError(err.Error())
+	}
+
+	return attrCfg, nil
+}
+
+// Validate creates the list of implicit dependencies.
+func (config *AttrConfig) Validate(path string) ([]string, error) {
+	if config.ConfigParams["mode"] == "" {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "config_params[mode]")
+	}
+
+	if config.DataDirectory == "" {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "data_dir")
+	}
+
+	if config.UseLiveData == nil {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "use_live_data")
+	}
+
+	if config.DataRateMsec < 0 {
+		return nil, errors.New("cannot specify data_rate_msec less than zero")
+	}
+
+	if config.MapRateSec != nil && *config.MapRateSec < 0 {
+		return nil, errors.New("cannot specify map_rate_sec less than zero")
+	}
+
+	deps := config.Sensors
+
+	return deps, nil
+}
+
+// SetOptionalParameters updates any unset optional config parameters to the values passed to this function.
+func (config *AttrConfig) SetOptionalParameters(port string, defaultDataRateMsec, defaultMapRateSec int, logger golog.Logger) error {
+	if config.Port == "" {
+		config.Port = port
+	}
```

Comment:
@zaporter-work  [nit/opt] I still get tripped up by the name vs. the fact that arguments are (only partially) prepended with `default...`. Can we rename `port` --> `defaultPort` to make it more consistent?

Reply:
If there's an argument not to, that's fine. It's really opt or super nit.

Reply:
Done. Good catch.